File: algo-anagrams-i/js/characterMatch.js
const isCharacterMatch = function (string1, string2) {
  const string1_obj = {};
  for (let letter of string1.toLowerCase()) {
    if (letter !== " ") {
      string1_obj[letter] = string1_obj[letter] + 1 || 1;
    }
  }
  for (let letter of string2.toLowerCase()) {
    if (letter !== " ") {
      if (letter in string1_obj) {
        string1_obj[letter] -= 1;
      } else {
        return false;
      }
    }
  }
  for (let letter in string1_obj) {
    if (string1_obj[letter] > 0) {
      return false;
    }
  }
  return true;
};

const anagramsFor = function (word, listOfWords) {
  const anagrams = [];
  for (let wordFromList of listOfWords) {
    if (isCharacterMatch(word, wordFromList)) {
      anagrams.push(wordFromList);
    }
  }
  return anagrams;
};

module.exports = { isCharacterMatch, anagramsFor };

File: algo-anagrams-i/python/character_match.py
# Part 1
def is_character_match(string1, string2):
    string1_dict = {}
    for letter in string1.lower():
        string1_dict[letter] = string1_dict.get(letter, 0) + 1
    for letter in string2.lower():
        if letter not in string1_dict:
            return False
        else:
            string1_dict[letter] -= 1
    for letter in string1_dict:
        if string1_dict[letter] > 0:
            return False
    return True


# Part 2
def anagrams_for(word, list_of_words):
    anagrams = []
    for word_in_list in list_of_words:
        if is_character_match(word, word_in_list):
            anagrams.append(word_in_list)
    return anagrams
